notes.md

```diff
@@ -0,0 +1,264 @@
+# Design Notes
+
+# Introduction
+
+This PR introduces the first working sketch of the django-pubsub library
+(still not sure about the name).
+It focuses on developing the beginnings of the user facing interface and basic
+mechanics of the python implementation of the postgres notify-listen protocol.
+We do not focus on the trickier topics of dealing with concurrency and
+listener durability.
+
+No design in here is by any means set in stone and very happy to clarify/change/discuss
+any aspect of this if there's a better idea out there. There are also a number
+of open questions/ideas for improvement listed throughout this doc/in comments
+in the code.
+
+I have tested as much as possible without having postgres to unit test over.
+There will still be a lot of bugs and probably code which could be refactored
+and improved.
+The examples below are actually working examples at least.
+
+
+# Documentation/Notes via Examples
+Here we show how a user can use django-pubsub to set up and notify listeners
+to process functions async. We'll make use of the models and relationships
+defined in the tests app.
+
+We focus on two use cases:
+- The first is where the user defines a listener with custom keyword arguments and notifies
+  that listener manually.
+- The second is where the user defines a listener which is notified whenever some django-pgtrigger is executed.
+
+Fully working versions of these examples can be found in the `tests` app.
+
+## Custom Listen and Manual Notify
+
+In this example, we wish to maintain a cache of how frequently
+Post objects are read per day. We want to defer this caching logic
+to another process so as not to hurt read performance times for posts.
+The user achieves this by first implementing a listener for the caching:
+
+Note: This example is for illustrative purposes only. There are of course
+many other ways to achieve this.
+
+
+### Declaring a Listener
+First, define the listener in their app's `tests.listeners.py` file:
+
+```python
+post_reads_per_date_cache = defaultdict(dict)  # dummy cache for illustrative purposes only
+
+@listen('post_reads_per_date')
+def post_reads_per_date(
+        *,
+        post_id: int,
+        date: datetime.date,
+):
+    print(f'Processing post_reads_per_date with args {post_id}, {date}')
+    print(f'Cache before: {post_reads_per_date_cache}')
+    current_count = post_reads_per_date_cache[date].get(post_id, 0)
+    post_reads_per_date_cache[date][post_id] = current_count + 1
+    print(f'Cache after: {post_reads_per_date_cache}')
+```
+
+The user must also ensure that this `listeners` module is imported into
+that app's config class (similar to signals):
+```python
+class TestsConfig(AppConfig):
+    name = 'tests'
+
+    def ready(self):
+        import tests.listeners
```

Comment:
Note - we can do this by default in a base app config provided by the library if we want, similar to how celery.py can autodiscover tasks.py. Happy to provide some sample code if you'd like

Reply:
Some sample code for this part could be nice :)